ext/introduction.asciidoc

```diff
@@ -121,7 +121,16 @@ The initial state of the compiler is as if the directive
 ----
 
 was issued, telling the compiler that all error and warning reporting must
-be done according to this specification, ignoring any extensions.
+be done according to the core specification, ignoring any extensions.
+
+Pragma directives are only required for extensions that change the compiler
+behavior i.e. the source code is interpreted differently if the pragma
+directive is used. Each extension should specify whether it adds the
+directive or not. However, if it is not explicitly specified it is assumed
+that the pragma directive is optional for the extension.  As unknown
```

Comment:
This PR updated the descriptions of `cl_khr_fp16` and `cl_khr_fp64` to indicate that they explicitly require pragma directives to enable the extensions.  Are there other extensions that also require pragma directives, even if they are vendor extensions?  If so, to the best of our ability I think we need to update all of the extensions that require a pragma, or we run the risk of breaking shipping implementations.

For example, it looks like the Intel extension `cl_intel_device_side_avc_motion_estimation` requires a pragma (rightly or wrongly): https://godbolt.org/z/T4TGYv

Alternatively, we could specify that an extension pragma is required by default unless it is explicitly specified to be optional.  This is more conservative, and I suspect that fewer extensions will require a pragma than extensions where the pragma is optional or not required, but this is also less likely to break shipping implementations.

Reply:
> Are there other extensions that also require pragma directives, even if they are vendor extensions? If so, to the best of our ability I think we need to update all of the extensions that require a pragma, or we run the risk of breaking shipping implementations.

Sure, I am happy to update those if we categorize them first.

> Alternatively, we could specify that an extension pragma is required by default unless it is explicitly specified to be optional. This is more conservative, and I suspect that fewer extensions will require a pragma than extensions where the pragma is optional or not required, but this is also less likely to break shipping implementations.

Sure. We can also change the default behavior. From the tooling implementation I would prefer to keep the default behavior to not require the pragma but accept it because:
1. Most of extensions won't need pragma in practice so not needing the pragma will likely be a common case.
2. We can simply implementation of many extensions straight away.

But it certainly has an implication on backward compatibility so if some vendors feel there is an issue to update we can fall back to the conservative approach.

Reply:
@bashbaug I have updated this PR without specifying the default behavior at all but however mandating each extension to provide the description of the pragma use.

The reason why I have changed it is that I personally think it is not sufficient to say whether the pragma is needed or not but it should also be explained what changes it makes. This is needed both for the compiler and application developers.

I have also added an explanation of pragma use in **cl_khr_fp64**. I have not added pragma to any other extension as I still doubt it is needed anywhere else. Note that I don't think guarding the use of types or functions that are available if an extension is supported by a pragma is useful and I have therefore excluded those (see my earlier comment).